common/debug-interface/src/trace.rs

```diff
@@ -0,0 +1,106 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+#![forbid(unsafe_code)]
+
+use crate::json_log::JsonLogEntry;
+use anyhow::Result;
+use chrono::{DateTime, Duration, Utc};
+use reqwest::Client;
+use serde_json::{json, Value};
+
+const TRACE_EVENT: &str = "trace_event";
+const TRACE_EDGE: &str = "trace_edge";
+const PAGING_SIZE: usize = 10000;
+
+pub struct LibraTraceClient {
+    client: Client,
+    addr: String,
+}
+
+impl LibraTraceClient {
+    /// Create LibraTraceClient from a valid socket address.
+    pub fn new<A: AsRef<str>>(address: A, port: u16) -> Self {
+        let client = Client::new();
+        let addr = format!("http://{}:{}", address.as_ref(), port);
+
+        Self { client, addr }
+    }
+
+    pub async fn get_libra_trace(
+        &self,
+        start_time: DateTime<Utc>,
+        duration: Duration,
+    ) -> Result<Vec<(String, JsonLogEntry)>> {
+        let start = start_time.format("%FT%T%.3fZ").to_string();
+        let end = (start_time + duration).format("%FT%T%.3fZ").to_string();
+        let response = self
+            .client
+            .get(&format!("{}/_search?scroll=1m", self.addr))
+            .json(&json!(
+            {
+                "size": PAGING_SIZE,
+                "query": {
+                    "bool": {
+                        "must": [
+                            { "term": { "name":   "libra_trace"}}
+                        ],
+                        "filter": [
+                            { "range": { "@timestamp": { "gte": start, "lte": end }}}
+                        ]
+                    }
+                }
+            }
+            ))
+            .send()
+            .await?;
+
+        let response: Value = response.json().await?;
+
+        let mut id = response["_scroll_id"].clone();
+        let mut v: Vec<(String, JsonLogEntry)> = Vec::new();
+        parse_response(&mut v, &response)?;
+
+        loop {
+            let response = self
+                .client
+                .get(&format!("{}/_search/scroll", self.addr))
+                .json(&json!(
+                {
+                    "scroll" : "1m",
+                    "scroll_id" : id
+                }
+                ))
+                .send()
+                .await?;
+            let response: Value = response.json().await?;
+            id = response["_scroll_id"].clone();
+            parse_response(&mut v, &response)?;
+            let hits = response["hits"]["hits"].as_array().unwrap().len();
+            if hits < PAGING_SIZE {
+                break;
+            }
+        }
+        Ok(v)
+    }
+}
+
+fn parse_response(v: &mut Vec<(String, JsonLogEntry)>, response: &Value) -> Result<()> {
```

Comment:
You need to change this function so it does not panic if something is wrong with response and return error instead.

One interesting option here is to use serde to actually deserialize response into some struct. For example, you can define
```
#[derive(Deserialize)]
struct Response {
hits: Hits
}
#[derive(Deserialize)]
struct Hits {
hits: Vec<Hit>
}
#[derive(Deserialize)]
struct Hit {
#[serde(rename="_source")]
source: Source;
}
#[derive(Deserialize)]
struct Source {
#[serde(rename="kubernetes.pod_name")]
pod_name: String,
data: serde_json::Value,
}
```

And after this you can use deserialization from json to verify response structure in one line